ocdata/core/surface.py

```diff
@@ -0,0 +1,275 @@
+    direction of the 3rd unit cell vector).
+
+    Arguments
+    ---------
+    surface_atoms: ase.Atoms
+
+    Returns
+    -------
+    tags: list
+        A list that contains the indices of the surface atoms.
+    """
+    unit_cell_height = np.linalg.norm(surface_atoms.cell[2])
+    scaled_positions = surface_atoms.get_scaled_positions()
+    scaled_max_height = max(scaled_position[2] for scaled_position in scaled_positions)
+    scaled_threshold = scaled_max_height - 2.0 / unit_cell_height
+
+    tags = [
+        0 if scaled_position[2] < scaled_threshold else 1
+        for scaled_position in scaled_positions
+    ]
+    return tags
+
+
+def find_surface_atoms_with_voronoi(bulk_atoms, surface_atoms):
+    """
+    Labels atoms as surface or bulk atoms according to their coordination
+    relative to their bulk structure. If an atom's coordination is less than it
+    normally is in a bulk, then we consider it a surface atom. We calculate the
+    coordination using pymatgen's Voronoi algorithms.
+
+    Note that if a single element has different sites within a bulk and these
+    sites have different coordinations, then we consider slab atoms
+    "under-coordinated" only if they are less coordinated than the most under
+    undercoordinated bulk atom. For example:  Say we have a bulk with two Cu
+    sites. One site has a coordination of 12 and another a coordination of 9.
+    If a slab atom has a coordination of 10, we will consider it a bulk atom.
+
+    Arguments
+    ---------
+    bulk_atoms: ase.Atoms
+        The bulk structure that the surface was cut from.
+    surface_atoms: ase.Atoms
+        The surface structure.
+
+    Returns
+    -------
+    tags: list
+        A list of 0s and 1s whose indices align with the atoms in
+        `surface_atoms`. 0s indicate a bulk atom and 1 indicates a surface atom.
+    """
+    # Initializations
+    surface_struct = AseAtomsAdaptor.get_structure(surface_atoms)
+    center_of_mass = calculate_center_of_mass(surface_struct)
+    bulk_cn_dict = calculate_coordination_of_bulk_atoms(bulk_atoms)
+    voronoi_nn = VoronoiNN(tol=0.1)  # 0.1 chosen for better detection
+
+    tags = []
+    for idx, site in enumerate(surface_struct):
+
+        # Tag as surface atom only if it's above the center of mass
+        if site.frac_coords[2] > center_of_mass[2]:
+            try:
+
+                # Tag as surface if atom is under-coordinated
+                cn = voronoi_nn.get_cn(surface_struct, idx, use_weights=True)
+                cn = round(cn, 5)
+                if cn < min(bulk_cn_dict[site.species_string]):
+                    tags.append(1)
+                else:
+                    tags.append(0)
+
+            # Tag as surface if we get a pathological error
+            except RuntimeError:
+                tags.append(1)
+
+        # Tag as bulk otherwise
+        else:
+            tags.append(0)
+    return tags
+
+
+def calculate_center_of_mass(struct):
```

Comment:
this is implemented in ase as `.get_center_of_mass()`

Reply:
Agreed. We can replace https://github.com/Open-Catalyst-Project/Open-Catalyst-Dataset/pull/62/files#diff-fab7f1bbdfa213d9fb154d05f5ed702baeaaab2f09506775227c60e206d30f02R204 with `center_of_mass = surface_atoms.get_center_of_mass()`.

Although I would sanity check this to make sure they're identical. The constant switching between the two has me a little nervous if for some reason pymatgen/ase reorders things differently, etc. I doubt it but I haven't looked into it...

Reply:
I think a sanity check is smart. Should be identical though!